nntrainer/graph/network_graph.cpp

```diff
@@ -0,0 +1,424 @@
+    return ML_ERROR_NONE;
+  }
+
+  if (num_node == 0) {
+    ml_loge("layer is empty");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (current.getType() == "activation") {
+    ml_loge("It is not allowed to realize ativation layer, possibly layer is "
+            "added right after activation");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (act == ActivationType::ACT_UNKNOWN) {
+    ml_loge("cannot realize unknown activation type");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  std::shared_ptr<Layer> layer = nntrainer::createLayer("activation");
+
+  ensureName(layer, current.getName());
+  layer->setActivation(act);
+  layer->num_inputs = 1;
+  layer->input_layers.clear();
+  layer->input_layers.push_back(current.getName());
+
+  addLayerNode(layer);
+
+  for (unsigned int i = 0; i < layers.size(); ++i) {
+    for (unsigned int j = 0; j < layers[i]->input_layers.size(); ++j) {
+      if (layers[i]->input_layers[j] == current.getName()) {
+        layers[i]->input_layers[j] = layer->getName();
+      }
+    }
+  }
+
+  return ML_ERROR_NONE;
+}
+
+int NetworkGraph::addLossLayer(const LossType loss_type) {
+  int status = ML_ERROR_NONE;
+  LossType updated_loss_type = loss_type;
+  if (num_node == 0) {
+    status = ML_ERROR_INVALID_PARAMETER;
+    NN_RETURN_STATUS();
+  }
+
+  if (updated_loss_type == LossType::LOSS_ENTROPY) {
+    if (getLayerNode(num_node - 1).layer->getType() != "activation") {
+      ml_loge("Error: Cross Entropy need last layer to have softmax or sigmoid "
+              "activation.");
+      return ML_ERROR_NOT_SUPPORTED;
+    }
+
+    LayerNode act_layer_node = getLayerNode(num_node - 1);
+    adj.pop_back();
+    num_node--;
+
+    switch (act_layer_node.layer->getActivationType()) {
+    case ActivationType::ACT_SIGMOID:
+      updated_loss_type = LossType::LOSS_ENTROPY_SIGMOID;
+      break;
+    case ActivationType::ACT_SOFTMAX:
+      updated_loss_type = LossType::LOSS_ENTROPY_SOFTMAX;
+      break;
+    default:
+      ml_loge("Error: Cross Entropy not supported without softmax or sigmoid.");
+      return ML_ERROR_NOT_SUPPORTED;
+    }
+  }
+
+  std::string input_str = getLayerNode(num_node - 1).layer->getName();
+
+  std::shared_ptr<Layer> layer = nntrainer::createLayer("loss");
+
+  ensureName(layer);
+
+  layer->num_inputs = 1;
+  layer->input_layers.clear();
+  layer->input_layers.push_back(input_str);
+
+  addLayerNode(layer);
+
+  return ML_ERROR_NONE;
+}
+
+void NetworkGraph::setOutputLayers(std::vector<std::shared_ptr<Layer>> layers) {
+
+  for (unsigned int idx = 0; idx < layers.size(); ++idx) {
+    unsigned int count = 0;
+    for (unsigned int i = 0; i < layers.size(); ++i) {
+      if (layers[i]->getName() == layers[idx]->getName())
+        continue;
+      for (unsigned int j = 0; j < layers[i]->input_layers.size(); ++j) {
+        if (layers[i]->input_layers[j] == layers[idx]->getName()) {
+          layers[idx]->output_layers.push_back(layers[i]->input_layers[j]);
```

Comment:
This sets the `layers[idx]->getName()` as the output layers of `layers[idx]` which I think is a bug.

Reply:
I don't think it is bug. Let's say ```a --> b --> c```. then the output of a is must be  b. and input of b must be a. So.. ```layers[0]->getName(): a == layer[1]->input_layers[0] : b --> true``` and set ```layers[0]->output_layers.push_back(layer[1]->input_layers[0] : b)```. then the output of a is b and input of b is a. I think everything set properly. I guess. Let me know if I'm wrong.

also ``` layers[idx]->output_layers.push_back(layers[i]<-- must be the string. )```

Reply:
This was fixed at 266bb3e. Resolving this issue.